### ISSUE TITLE
[Feature Request] Logging for 401 response and possible Json response

### ISSUE THREAD

**Issue by upMKuhn** [babelouest/glewlwyd#52]
Hi, 

I am currently battling with the Admin interface. All requests return 401, despite the fact that I am logged in.  `https://my.domain.com/api/user/?limit=10`. Can we add some logging providing a specific reason why and most importantly that it returned 401.

My token is valid and the signature checks out
> eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzUxMiJ9.eyJleHBpcmVzX2luIjo4MjAwLCJpYXQiOjE1NDk4NDM5OTksInNhbHQiOiImNTdBWUxaYWdXPChGQVY1IiwidHlwZSI6ImFjY2Vzc190b2tlbiIsInVzZXJuYW1lIjoiYWRtaW4ifQ.abcd1234

By the way, I am using the latest release. The current quickstart docker image works perfectly. :(  

Thanks for creating the project by the way :)

**Comment by babelouest** [comment]
Hello,

This feature request is already in my todo list for the [next release](https://github.com/babelouest/glewlwyd/tree/2.0b), but it is not ready yet though :-p

Even if I'm missing some context, I assume you're talking about the API requests that return an error 401, even if you're logged in.

If I read your jwt, I don't see any scope value in it, which means that you have set this value in your config file:
> use_scope=false

I tried and installed a fresh Glewlwyd with the default settings and `use_scope=false` and I didn't have any problem on my side.
So I'll have some difficulties to reproduce your bug.

Have you installed Glewlwyd as described in the [INSTALL.md](https://github.com/babelouest/glewlwyd/blob/master/docs/INSTALL.md) file? If you did something that isn't described in this doc or if you made a change from it, can you describe what change?

If possible, can you post your configuration file **WITHOUT** the passwords, secrets or any other sensitive data? I may have better luck helping you with that.

**Comment by upMKuhn** [comment]
Thanks, for looking into it ;) 
Yes, most likely it is connected to either the DB or the config file. :( 
Regarding the auth token not containing any scope. I had it also set to true yesterday, but it was the same result. I used the JWT debugger and made sure it contained the two scopes. g_admin and g_profile. The signature of the RSA key also checked out.  

My current config
```
# port to open for remote commands
port=4593
# url prefix
url_prefix="api"
# path to static files for /app url
static_files_path="/usr/share/glewlwyd/webapp/"
# static files prefix
static_files_prefix = "app"
# mime types for static files
static_files_mime_types =
(
    {
    extension = ".html"
    type = "text/html"
    },
    {
    extension = ".css"
    type = "text/css"
    },
    {
    extension = ".js"
    type = "application/javascript"
    },
    {
    extension = ".png"
    type = "image/png"
    },
    {
    extension = ".jpg"
    type = "image/jpeg"
    },
    {
    extension = ".jpeg"
    type = "image/jpeg"
    },
    {
    extension = ".ttf"
    type = "font/ttf"
    },
    {
    extension = ".woff"
    type = "font/woff"
    },
    {
    extension = ".woff2"
    type = "font/woff2"
    },
    {
    extension = ".map"
    type = "application/octet-stream"
    },
    {
    extension = ".ico"
    type = "image/x-icon"
    }
)
# access-control-allow-origin value
allow_origin="*"
# log mode (console, syslog, file)
log_mode="console"
# log level: NONE, ERROR, WARNING, INFO, DEBUG
log_level="DEBUG"
# output to log file (required if log_mode is file)
log_file="/tmp/glewlwyd.log"
# use scope for access control (default true)
use_scope=true
# refresh token expiration in seconds, default is 2 weeks
refresh_token_expiation=1209600
# access token expiration in seconds, default is 1 hour
access_token_expiration=8200
# session expiration, default is 4 weeks
session_expiration=2419200
# session key
session_key="GLEWLWYD_SESSION_ID"
# admin scope name
admin_scope="g_admin"
# profile scope name
profile_scope="g_profile"
# additional property name that will be included in access tokens, uncomment to enable this feature
# additional_property_name  = "new_property"
# code expiration in seconds, default is 10 minutes
code_expiration=600
# TLS/SSL configuration values
use_secure_connection=false
secure_connection_key_file="/usr/etc/glewlwyd/cert.key"
secure_connection_pem_file="/usr/etc/glewlwyd/cert.pem"
# login and grant urls
login_url="https://domain.com/app/login.html?"
grant_url="https://domain.com/app/grant.html?"
# Reset password configuration
reset_password=false # optional, default false
reset_password_config = 
{
# SMTP parameters
    smtp_host="mail.domain.com"    # mandatory
    smtp_port="587"               # optional, default 25
    smtp_use_tls=true             # optional, default false
    smtp_verify_certificate=true # optional, default false
    smtp_user="root@domain.com"             # optional, default no value
    smtp_password="password"     # optional, default no value
    
    token_expiration=8200                                                                       # in seconds, optional, default 7 days
    email_from="root@domain.com"                                                           # mandatory
    email_subject="Kuhn's Password reset"                                                     # mandatory
    email_template="reset.html"                                                                  # mandatory
    page_url_prefix="https://domain.com/glewlwyd/app/reset.html?user=$USERNAME&code=$TOKEN"    # mandatory
}
# Algorithms available are SHA1, SHA256, SHA512, MD5, default is SHA256
hash_algorithm = "SHA256"
# MariaDB/Mysql database connection
database =
{
    type               = "sqlite3";
    path               = "/var/cache/glewlwyd/glewlwyd.db";
};
# Authentication configuration
authentication =
{
    database_auth = true
}
# jwt parameters
jwt =
{
    # key size for algorithms, values available are 256, 384 or 512, default 512
    key_size = 512
    
    # Use RSA algorithm to sign tokens (asymetric)
    use_rsa = true
    
    # path to the key (private) certificate file to sign tokens
    rsa_key_file = "/var/glewlwyd/keys/rsa.key"
    
    # path to the public certificate file to validate signatures
    rsa_pub_file = "/var/glewlwyd/keys/rsa.pem"

    # Use SHA algorithm to sign tokens (symetric)
    use_sha = false
   
    # characters used to generate and validate the token
    sha_secret = "secret"
    
}
```

**Comment by babelouest** [comment]
Hello @upMKuhn ,

If I use your config file I have a normal working glewlwyd, so the problem comes from elsewhere.

First, check that you have initialized the database with the proper script. Since you use `hash_algorithm = "SHA256"` and `sqlite3` database, you must use the script [init-sqlite3-sha256.sql](https://github.com/babelouest/glewlwyd/blob/master/docs/database/init-sqlite3-sha256.sql)

Also, there may be a problem with your RSA files, although this should have been detected on startup. But you can try to re-generate the key/pem files or use sha to test if the problem remains.

If none of this works, we'll try something else.

**Comment by upMKuhn** [comment]
I ran just now the entire setup with the same config, rsa keys and image in docker compose. It worked there perfectly.

Thanks for the effort. I think I managed to narrow down the problem. I might have failed to mention, that I am running your app in my kubernetes "cluster". After reviewing the authentication.c file, I stumbled across the config option "auth_code_match_ip_address". Could that be of relevance? The pod is running on an overlay network, which hides the true source ip.
I tried `auth_code_match_ip_address=false`, but it didn't do much :(

**Comment by babelouest** [comment]
The `auth_code_match_ip_address=false` is used only when you require a code for a refresh token in response type `code`. But that's not relevant for your case.

If you're running glewlwyd on a kubernetes cluster, this means that you have several instances running at the same time.
Do you use the same rsa keys for all instances? Maybe that's your problem in the first place, especially if you use Docker images that generate rsa keys on first run. In that case you'll have all docker instances running with a different key, therefore an error 401 most of the time.

You can share the same config file and rsa keys with multiple instances of a glewlwyd. But I wouldn't recommend using SQLite3 database in a kubernetes cluster because the database stores codes and refresh keys signatures so one instance will not know what the next one has generated. Although sharing a mysql database between all instances should be ok.

To summarize, my recommendations for you:
- Export the volume `/var/glewlwyd` on the Docker instances to a shared folder that will be the same for all Docker instances
- Use a calibrated MySQL database for all the Docker instances so they will share the same context.

**Comment by upMKuhn** [comment]
I have limted the replication to 1 pod, bound it to a single node and all volumes are on the node's file system. I know a bit contradictory against the way of kubernetes, but for my home lab with limited resources it is working fine.  Regarding access to the rsa keys. I did ssh-ed into the container and copied content for my docker-compose replication yestderday. I also deleted the db file and let my container generate it again on both the cluster and local.  

When I get home I'll check that all headers are forwarded, maybe the nginx reverse proxy strips the headers accidentally. (This would be weird though, because I can go on to the profile page without any issues.)

**Comment by babelouest** [comment]
Glewlwyd is not designed for concurrent access on SQLite3 databases, that's what Mysql databases are for.
As I said, if you use glewlwyd in a cluster, the only way to have consistent behaviour is to use Mysql backend, not SQLite3.

The profile page relies on the session cookie, not on the access_token, that's probably why you can make it work.

Let me know if you find something with the nginx proxy (I don't know this one, I only use apache mod proxy_http).

**Comment by upMKuhn** [comment]
Ok, I finally made it work. My nginx proxy is configured to do SSL termination. (because its refreshing my certificates automatically)    

I believe there seems to be an issue with the client lower casing the header name Authorization when *using* HTTPS. I used postman and an [nginx debug image](https://hub.docker.com/r/brndnmtthws/nginx-echo-headers/dockerfile) to proof that the headers can arrive in the same case as it was sent. Postman was able to get a 200 response with the exact casing `Authorization`, while using HTTPS.  I know it sounds weird that the case sensitivity changes over https.

### The logs:

Postman
```
GET /api/user/ HTTP/1.1
Host: host
X-Request-ID: 5632add1283739819982a1ce9fe25b37
X-Real-IP: 10.244.0.0
X-Forwarded-For: 10.244.0.0
X-Forwarded-Port: 443
X-Forwarded-Proto: https
X-Original-URI: /api/user/
X-Scheme: https
AUThorization: a
cache-control: no-cache
Postman-Token: 68d5a1b4-3823-42f6-b5bc-20cee0a9cf19
User-Agent: PostmanRuntime/7.6.0
Accept: */*
accept-encoding: gzip, deflate
```
Firefox:
```
GET /api/user/?limit=10 HTTP/1.1
Host: auth.kuhn.host
X-Request-ID: 2a4f8efa1ab664674cf7c086b6902b76
X-Real-IP: 10.244.0.0
X-Forwarded-For: 10.244.0.0
X-Forwarded-Host: domain
X-Forwarded-Port: 443
X-Forwarded-Proto: https
X-Original-URI: /api/user/?limit=10
X-Scheme: https
user-agent: Mozilla/5.0 (X11; Linux x86_64; rv:65.0) Gecko/20100101 Firefox/65.0
accept: */*
accept-language: en-US,en;q=0.5
accept-encoding: gzip, deflate, br
referer: https://domain/app/
authorization: Bearer ...
x-requested-with: XMLHttpRequest
cookie: ....
```

Regarding my choice of database. I use the system for my personal setup, learning purposes. I only require one instance running at a time. The load should be relatively low. I need to save my system resources to pack more stuff on one node :grin:

**Comment by babelouest** [comment]
Awesome!

As I said, I'm not fluent in nginx so I trust you with your solution.

**Comment by upMKuhn** [comment]
Can you please fix the backend's case sensivity :) 

Failing request 
```
curl -i -H "authorization: Bearer myToken" http://localhost/api/user/\?limit\=10 
```

Passing request 
```
curl -i -H "Authorization: Bearer myToken" http://localhost/api/user/\?limit\=10 
```
Thanks :) 

https://stackoverflow.com/questions/5258977/are-http-headers-case-sensitive

**Comment by babelouest** [comment]
Glewlwyd 1.4 is currently on hold because I'm working on the [branch 2.0](https://github.com/babelouest/glewlwyd/tree/2.0b). I'll maintain the branch 1.4 for bug fixes and security issues but not for new features.

I'm not comfortable with changing a code that works correctly because a third-party program changes its behavior.

I will make this change for future release (>= 2.0), but I won't make the change in the current branch 1.4.
Nevertheless, here is a patch you can apply to your code. This should fix your problem, but it's without any guarantee of success :p

Spoiler alert: The patch replaces all `u_map_get(request->map_header, "Authorization")` with `u_map_get_case(request->map_header, "Authorization")`.

```patch
diff --git a/src/webservice.c b/src/webservice.c
index 3a0e053..d0d41dc 100644
--- a/src/webservice.c
+++ b/src/webservice.c
@@ -182,7 +182,7 @@ int callback_glewlwyd_set_user_scope_grant (const struct _u_request * request, s
   int res;
   json_t * j_scope, * j_session;
   
-  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get(request->map_header, "Authorization"));
+  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get_case(request->map_header, "Authorization"));
   j_scope = auth_check_user_scope(((struct config_elements *)user_data), json_string_value(json_object_get(json_object_get(j_session, "grants"), "username")), u_map_get(request->map_post_body, "scope"));
   if (!check_result_value(j_scope, G_OK)) {
     response->status = 403;
@@ -324,7 +324,7 @@ int callback_glewlwyd_check_user (const struct _u_request * request, struct _u_r
   json_t * j_auth = NULL;
   int res = U_CALLBACK_CONTINUE;
   
-  j_auth = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get(request->map_header, "Authorization"));
+  j_auth = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get_case(request->map_header, "Authorization"));
   if (!check_result_value(j_auth, G_OK)) {
     res = U_CALLBACK_UNAUTHORIZED;
   } else {
@@ -360,7 +360,7 @@ int callback_glewlwyd_check_scope_admin (const struct _u_request * request, stru
   json_t * j_session = NULL;
   int res = U_CALLBACK_UNAUTHORIZED;
   
-  j_session = access_token_check_scope_admin(config, u_map_get(request->map_header, "Authorization"));
+  j_session = access_token_check_scope_admin(config, u_map_get_case(request->map_header, "Authorization"));
   if (check_result_value(j_session, G_OK)) {
     res = U_CALLBACK_CONTINUE;
   }
@@ -375,12 +375,12 @@ int callback_glewlwyd_get_user_session_profile (const struct _u_request * reques
   struct config_elements * config = (struct config_elements *)user_data;
   json_t * j_session = NULL, * j_user = NULL, * j_grant = NULL;
   
-  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get(request->map_header, "Authorization"));
+  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get_case(request->map_header, "Authorization"));
   if (check_result_value(j_session, G_OK)) {
     if (json_object_get(json_object_get(j_session, "grants"), "session_token") == json_true()) {
       j_user = get_user(config, json_string_value(json_object_get(json_object_get(j_session, "grants"), "username")), NULL);
     } else {
-      j_grant = access_token_check_scope_profile(config, u_map_get(request->map_header, "Authorization"));
+      j_grant = access_token_check_scope_profile(config, u_map_get_case(request->map_header, "Authorization"));
       j_user = get_user_profile(config, json_string_value(json_object_get(json_object_get(j_session, "grants"), "username")), NULL, json_string_value(json_object_get(json_object_get(j_grant, "grants"), "scope")));
       json_decref(j_grant);
     }
@@ -403,7 +403,7 @@ int callback_glewlwyd_delete_user_session (const struct _u_request * request, st
   json_t * j_session;
   
   if (u_map_get(request->map_cookie, config->session_key) != NULL && strlen(u_map_get(request->map_cookie, config->session_key)) > 0) {
-    j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get(request->map_header, "Authorization"));
+    j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get_case(request->map_header, "Authorization"));
     if (check_result_value(j_session, G_OK)) {
       session_hash = generate_hash(config, config->hash_algorithm, u_map_get(request->map_cookie, config->session_key));
       if (revoke_session(config, json_string_value(json_object_get(json_object_get(j_session, "grants"), "username")), session_hash) != G_OK) {
@@ -423,7 +423,7 @@ int callback_glewlwyd_delete_user_session (const struct _u_request * request, st
  */
 int callback_glewlwyd_get_user_session_scope_grant (const struct _u_request * request, struct _u_response * response, void * user_data) {
   struct config_elements * config = (struct config_elements *)user_data;
-  json_t * j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get(request->map_header, "Authorization"));
+  json_t * j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get_case(request->map_header, "Authorization"));
   json_t * j_scope_grant;
   
   if (!check_result_value(j_session, G_OK)) {
@@ -448,7 +448,7 @@ int callback_glewlwyd_user_scope_delete (const struct _u_request * request, stru
   json_t * j_scope, * j_session;
   
   // Check if user has access to scopes
-  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get(request->map_header, "Authorization"));
+  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get_case(request->map_header, "Authorization"));
   j_scope = auth_check_user_scope(((struct config_elements *)user_data), json_string_value(json_object_get(json_object_get(j_session, "grants"), "username")), u_map_get(request->map_post_body, "scope"));
   if (!check_result_value(j_scope, G_OK)) {
     response->status = 403;
@@ -1139,7 +1139,7 @@ int callback_glewlwyd_set_user_profile (const struct _u_request * request, struc
   struct config_elements * config = (struct config_elements *)user_data;
   json_t * j_session = NULL, * j_user_valid = NULL;
 
-  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get(request->map_header, "Authorization"));
+  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get_case(request->map_header, "Authorization"));
   if (check_result_value(j_session, G_OK)) {
     json_t * json_body = ulfius_get_json_body_request(request, NULL);
     j_user_valid = is_user_profile_valid(config, json_string_value(json_object_get(json_object_get(j_session, "grants"), "username")), json_body);
@@ -1258,7 +1258,7 @@ int callback_glewlwyd_get_refresh_token_profile (const struct _u_request * reque
     }
   }
   
-  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get(request->map_header, "Authorization"));
+  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get_case(request->map_header, "Authorization"));
   if (check_result_value(j_session, G_OK)) {
     if (u_map_get(request->map_url, "offset") != NULL) {
       offset = strtol(u_map_get(request->map_url, "offset"), NULL, 10);
@@ -1294,7 +1294,7 @@ int callback_glewlwyd_delete_refresh_token_profile (const struct _u_request * re
   json_t * j_session;
   int res;
   
-  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get(request->map_header, "Authorization"));
+  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get_case(request->map_header, "Authorization"));
   if (check_result_value(j_session, G_OK)) {
     json_t * json_body = ulfius_get_json_body_request(request, NULL);
     res = revoke_token(config, json_string_value(json_object_get(json_object_get(j_session, "grants"), "username")), json_string_value(json_object_get(json_body, "token_hash")));
@@ -1332,7 +1332,7 @@ int callback_glewlwyd_get_session_profile (const struct _u_request * request, st
     }
   }
   
-  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get(request->map_header, "Authorization"));
+  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get_case(request->map_header, "Authorization"));
   if (check_result_value(j_session, G_OK)) {
     if (u_map_get(request->map_url, "offset") != NULL) {
       offset = strtol(u_map_get(request->map_url, "offset"), NULL, 10);
@@ -1368,7 +1368,7 @@ int callback_glewlwyd_delete_session_profile (const struct _u_request * request,
   json_t * j_session, * j_body = ulfius_get_json_body_request(request, NULL);
   int res;
   
-  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get(request->map_header, "Authorization"));
+  j_session = session_or_access_token_check(config, u_map_get(request->map_cookie, config->session_key), u_map_get_case(request->map_header, "Authorization"));
   if (check_result_value(j_session, G_OK)) {
     json_t * json_body = ulfius_get_json_body_request(request, NULL);
     res = revoke_session(config, json_string_value(json_object_get(json_object_get(j_session, "grants"), "username")), json_string_value(json_object_get(json_body, "session_hash")));
```

**Comment by upMKuhn** [comment]
Awesome thanks :+1: 

I understand that pushing out a rushed bug fix release isn't the best idea. Happened to me more than once to make things worse :grinning:  Just for reference sake, the HTTP headers are by IETF standard case in-sensitive. I didn't know that as well till now ;)

**Comment by babelouest** [comment]
I knew that header names are case insensitive according to the RFC. The problem I have with that is that the `Authorization` header is a very important header that bears the security token.

Since not everyone is full RFC compliant, other users might have build resource services that rely on Glewlwyd's access tokens with the header called `"Authorization"` case sensitive.

So to avoid security issues on existing instances of Glewlwyd and having users angry at me, I can't change this in a minor release if it's not a bug.

You never know if you won't change someone's workflow that used to worked perfectly.
Other services rely on Glewlwyd for security purposes.

**Comment by upMKuhn** [comment]
:confused: Case-sensitivity means in my book headers spelled auThoriZatION | authorization | Authorization should be accepted as well. Anyway as I said, I understand that you don't want to do rushed  release, to avoid breaking things. Been there.

Thanks, 
for your help :)
